skipcpio/skipcpio.c

```diff
@@ -60,59 +90,78 @@ int main(int argc, char **argv)
         fseek(f, 0, SEEK_SET);
 
         /* check, if this is a cpio archive */
-        if (buf[0] == '0' && buf[1] == '7' && buf[2] == '0' && buf[3] == '7' && buf[4] == '0' && buf[5] == '1') {
+        if (memcmp(buf.cpio.h.c_magic, CPIO_MAGIC, 6) == 0) {
+
                 long pos = 0;
 
-                /* Search for CPIO_END */
-                do {
-                        char *h;
-                        fseek(f, pos, SEEK_SET);
-                        buf[sizeof(buf) - 1] = 0;
-                        s = fread(buf, CPIO_ENDLEN, 2, f);
-                        if (s <= 0)
-                                break;
+                unsigned long filesize;
+                unsigned long filename_length;
 
-                        h = memmem(buf, sizeof(buf), CPIO_END, sizeof(CPIO_END));
-                        if (h) {
-                                pos = (h - buf) + pos + CPIO_ENDLEN;
+                do {
+                        // zero string, spilling into next unused field, to use strtol
+                        buf.cpio.h.c_chksum[0] = 0;
+                        filename_length = strtol(buf.cpio.h.c_namesize, NULL, 16);
+                        pos = ALIGN_UP(pos + sizeof(struct cpio_header) + filename_length, CPIO_ALIGNMENT);
+
+                        // zero string, spilling into next unused field, to use strtol
+                        buf.cpio.h.c_dev_maj[0] = 0;
+                        filesize = strtol(buf.cpio.h.c_filesize, NULL, 16);
```

Comment:
This should be `strtoul()` (same for comment above). looks fine otherwise.

Reply:
you are so right.. thanks

Reply:
done